### PR TITLE
Added configuration with consul and flags

### PR DESCRIPTION
**Pull request by ericm**
Will run readError in its own goroutine maybe so that it can update config regularly

### CODE REVIEW

**Comment by ericm** [comment]
Works with dev env currently. Not 100% sure how consul auth is setup in production

**Comment by Strum355** [comment]
what does the KV look like in consul?

File: template.env [hunk 1]
```diff
@@ -1 +1,2 @@
-DISCORD_TOKEN=
\ No newline at end of file
+DISCORD_TOKEN=
+CONSUL_ADDRESS=172.17.0.1:8500
```
Comment (Strum355) [inline template.env:2]: for dev-env, this can just be `consul` because theyre in the same network namespace

**Comment by ericm** [comment]
image
mapping to strings of the server ids

File: template.env (hunk 1 above)
Comment (ericm) [reply template.env:2]: Oh right

**Comment by Strum355** [comment]
You could have both under a common key, as the value could be a json object containing the two values. Just an idea, as the value can be anything, even JSON

Either way, lowercase keys would be preferred imo

File: main.go [hunk 2]
```diff
@@ -1,42 +1,121 @@
...
+		log.InitSimpleLogger(&log.Config{Output: os.Stdout})
 	}
-	session, err := discordgo.New("Bot " + token)
-	if err != nil {
-		panic("Couldn't init discord connection")
+
+	// Consul config
+	consulConfig := api.DefaultConfig()
+	if production {
+		consulConfig.Scheme = "https"
```
Comment (Strum355) [inline main.go:40]: this will always be http, as theyre in a private network namespace

**Comment by ericm** [comment]
Oh right. That would be more maintainable

**Comment by ericm** [comment]
image

**Comment by Strum355** [comment]
*committee :)

File: main.go [hunk 3]
```diff
@@ -1,42 +1,121 @@
...
+	// Listen for consul updates
+	params := map[string]interface{}{
+		"type": "key",
+		"key":  "discordbot/servers",
+	}
+	watch, err := consulwatch.Parse(params)
+	exitError(err)
+	watch.Handler = func(idx uint64, data interface{}) {
+		// Serialize data
+		buf, err := json.MarshalIndent(data, "", "    ")
```
Comment (Strum355) [inline main.go:77]: what does `data` look like here? can you do `fmt.Printf("%#v %T", data, data)`

Comment (ericm) [reply main.go:77]: `&api.KVPair{Key:"discordbot/servers", CreateIndex:0xb, ModifyIndex:0xb, LockIndex:0x0, Flags:0x0, Value:[]uint8{0x7b, 0x22, 0x63, 0x6f, 0x6d, 0x6d, 0x69, 0x74, 0x74, 0x65, 0x65, 0x22, 0x3a, 0x20, 0x22, 0x73, 0x73, 0x22, 0x2c, 0x20, 0x22, 0x70, 0x75, 0x62, 0x6c, 0x69, 0x63, 0x22, 0x3a, 0x20, 0x22, 0x73, 0x73, 0x73, 0x22, 0x7d}, Session:"", Namespace:""} *api.KVPair`

Comment (ericm) [reply main.go:77]: Ill change the parser

Comment (Strum355) [reply main.go:77]: you can do `kv := data.(*api.KVPair)` and then just do `kv.Value` and happy days

File: main.go [hunk 4]
```diff
@@ -1,42 +1,111 @@
...
 	<-sc
-	fmt.Println("Cleanly exiting")
+	log.Info("Cleanly exiting")
 	session.Close()
 }
+
+func readConfig(consulConfig *api.Config) {
+	// Connect to consul
+	// Listen for consul updates
+	params := map[string]interface{}{
```
Comment (Strum355) [inline main.go:68]: https://github.com/hashicorp/consul/blob/master/command/watch/watch.go#L116-L118 We would need a field for the consul token, which would come from env :)

File: main.go [hunk 5]
```diff
@@ -1,42 +1,111 @@
...
+			return
+		}
+		if len(structData.Value) == 0 {
+			log.Error("servers key doesnt exist")
+			return
+		}
+		// Decode base64 Value
+		if err != nil {
+			log.Error(err.Error())
+		}
```
Comment (Strum355) [inline main.go:87]: Remnants from base64?

File: main.go (hunk 4 above)
Comment (Strum355) [reply main.go:69]: I dont believe it needs to be limited to production. Empty string is valid

**Comment by ericm** [comment]
@Strum355 any other changes?

**Pull request merged**